martha/martha_v3.js

```diff
@@ -407,30 +469,31 @@ async function retrieveFromServers(params) {
         localizationPath = getLocalizationPath(drsProvider, drsResponse);
 
         try {
-            let accessToken;
-            if (drsProvider.shouldFetchFenceAccessToken(accessMethod, requestedFields)) {
-                try {
-                    const bondAccessTokenUrl = `${config.bondBaseUrl}/api/link/v1/${bondProvider}/accesstoken`;
-                    console.log(`Requesting Bond access token for '${url}' from '${bondAccessTokenUrl}'`);
-                    const accessTokenResponse = await apiAdapter.getJsonFrom(bondAccessTokenUrl, auth);
-                    accessToken = accessTokenResponse.token;
-                } catch (error) {
-                    throw new RemoteServerError(error, 'Received error contacting Bond.');
-                }
-            }
+            const accessToken = await maybeFetchFenceAccessToken(accessMethod);
 
             // Retrieve the accessUrl using the returned accessToken, even if the token was empty.
             if (drsProvider.shouldFetchAccessUrl(accessMethod, requestedFields)) {
-                try {
-                    const httpsAccessUrl = generateAccessUrl(drsProvider, urlParts, accessMethod.access_id);
-                    // Use the access token fetched in the call above or the auth submitted to Martha directly by the
-                    // caller as appropriate.
-                    const accessUrlAuth = drsProvider.determineAccessUrlAuth(accessMethod, accessToken, auth);
-                    console.log(`Requesting DRS access URL for '${url}' from '${httpsAccessUrl}'`);
-                    accessUrl = await apiAdapter.getJsonFrom(httpsAccessUrl, accessUrlAuth);
-                } catch (error) {
-                    throw new RemoteServerError(error, 'Received error contacting DRS provider.');
+                const fetchAccessUrl = async () => {
```

Comment:
I made this a function because repeated assignments of `accessUrl`, especially one inside an if block whose condition referenced `accessUrl` made the linter complain about race conditions.